designs/disruption-controls-by-reason.md

```diff
@@ -0,0 +1,312 @@
+
+## User Scenarios 
+1. Users need the capability to schedule upgrades only during business hours or within more restricted time windows. Additionally, they require a system that doesn't compromise the cost savings from consolidation when upgrades are blocked due to drift.
+2. Users want to minimize workload disruptions during business hours but still want to be able to delete empty nodes throughout the day.  That is, empty can run all day, while limiting cost savings and upgrades due to drift to non-business hours only.
+
+See Less Made Up Scenarios here: 
+- https://github.com/kubernetes-sigs/karpenter/issues/924 
+- https://github.com/kubernetes-sigs/karpenter/issues/753#issuecomment-1790110838
+- https://github.com/kubernetes-sigs/karpenter/issues/672
+
+## Clarifying the requirements and behavior 
+**Reason and Budget Definition:** Users should be able to define an reason and a corresponding budget(s).
+**Supported Reasons:** All disruption Reasons affected by the current Budgets implementation (underutilized, empty, expired, drifted) should be supported. 
+**Default Behavior for Unspecified Reasons:** Budgets should continue to support a default behavior for all disruption reasons. 
+
+
+### Q: How should Karpenter handle the default or undefined reason case? 
+If a budget reason is unspecified like budgets[2], we will assume this budget applies to all actions that are not specified 
+```yaml
+budgets: 
+  - nodes: 10
+    reasons: [drifted, underutilized]
+    schedule: "* * * * *"
+  - nodes: 30 
+    schedule: "* * * * *" 
+```
+Meaning that for any actions other than drifted + consolidation, the total amount of disrupting + unhealthy nodes has to be less than 30 for them to trigger disruption.
+
+### Q: How should allowed disruptions be calculated in a multi-reason world? 
+The calculation of allowed disruptions in a system with multiple disruption reasons has become more intricate following the introduction of Disruption Budget Reasons. Previously, the formula was straightforward:
+
+AllowedDisruptions = minNodeCountOfAnActiveBudget - unhealthyNodes - totalDisruptingNodes.
+
+When calculating by reason, two potential equations emerge:
+1. AllowedDisruptionByReason = minNodeCountOfActiveBudget[reason] - unhealthyNodes - totalDisruptingNodes[reason]
+2. AllowedDisruptionByReason = minNodeCountOfActiveBudget[reason] - unhealthyNodes - totalDisruptingNodes
+The second equation is the one we should opt into.
+
+Take this budget as an example
+```yaml
+budgets: 
+  - nodes: 15 
+    reasons: [drifted, underutilized]
+    schedule: "* * * * *"
+  - nodes: 10
+    reasons: [drifted]
+    schedule: "* * * * *"
+  - nodes: 5 
+    schedule: "* * * * *" 
+```
+First, calculate minNodeCountOfActiveBudget. Given there are two budgets for drifted, we select the lower number:
+
+```
+minNodeCountOfActiveBudget = {
+  drifted: 10 
+  underutilized: 15 
+  Default: 5
```

Comment:
one question on the explanation here:

In this specific example, the nodes: 5 will always be the budget, since it's always enabled, and effectively is the same as `reasons: [*]`. Can we change up the example here with maybe changing this to a percent so it's possible to see that all the different budgets can be active at different points?

Reply:
+1, I think I had a similar question up above where we are making the determination on what not specifying the reason means.

Reply:
>  , and effectively is the same as reasons: [*] 
"" != "all" its all that are undefined was my intention. Do you expect that if I have 


minNodeCountOfActiveBudget = {
  drifted: 10 
  underutilized: 15 
  Default: 5
 }
 
 That if the 5 Node declaration should bound ALL disruption reasons?  That it should floor all reasons? That equation is slightly different from what I am proposing.